tools/linter/adapters/clickhouse_sql_linter.py

```diff
@@ -0,0 +1,172 @@
+import argparse
+import concurrent.futures
+import json
+import logging
+import os
+import re
+import subprocess
+import time
+from enum import Enum
+from typing import List, NamedTuple, Optional, Pattern
+
+
+LINTER_CODE = "CLICKHOUSE"
+
+
+class LintSeverity(str, Enum):
+    ERROR = "error"
+    WARNING = "warning"
+    ADVICE = "advice"
+    DISABLED = "disabled"
+
+
+class LintMessage(NamedTuple):
+    path: Optional[str]
+    line: Optional[int]
+    char: Optional[int]
+    code: str
+    severity: LintSeverity
+    name: str
+    original: Optional[str]
+    replacement: Optional[str]
+    description: Optional[str]
+
+
+RESULTS_RE: Pattern[str] = re.compile(
+    r"""(?mx)
+    ^
+    (?P<file>.*?):
+    (?P<line>\d+):
+    (?P<char>\d+):
+    \s(?P<message>.*)
+    \s(?P<code>\[.*\])
+    $
+    """
+)
+
+
+def run_command(
+    args: List[str],
+) -> "subprocess.CompletedProcess[bytes]":
+    logging.debug("$ %s", " ".join(args))
+    start_time = time.monotonic()
+    try:
+        return subprocess.run(
+            args,
+            capture_output=True,
+        )
+    finally:
+        end_time = time.monotonic()
+        logging.debug("took %dms", (end_time - start_time) * 1000)
+
+
+def check_file(
+    binary: str,
+    filename: str,
+) -> List[LintMessage]:
+    with open(filename) as f:
+        original = f.read()
+
+    try:
+        proc = run_command(
```

Comment:
This command removes the comments from the query.  Let me check https://github.com/ClickHouse/ClickHouse for more details.  The doc says there is `--comments` to keep the comments but it doesn't seem to work for me https://clickhouse.com/docs/en/operations/utilities/clickhouse-format

Reply:
After some digging, I think we should try https://github.com/sqlfluff/sqlfluff instead.  The fact that ClickHouse official binary removes all the comments in the query is a deal breaker IMO.  I guess it just isn't built for linting purpose in mind.  When it comes to linting, https://github.com/sqlfluff/sqlfluff does a much better job.  Not only it can autoformat the query (like black, ufmt, ruff), it also provides other SQL linters that we could explore further https://github.com/sqlfluff/sqlfluff/tree/main/src/sqlfluff/rules

https://github.com/pytorch/test-infra/issues/5837#issuecomment-2603112211